### ISSUE TITLE
pylint does not support typing.Self when override

### ISSUE THREAD

**Issue by alonbl** [pylint-dev/pylint#9159]
### Bug description

The following code in python-3.11 and in previous python using `typing_extensions`:

```python
import typing

class A:
    @classmethod
    def f1(cls) -> typing.Self:
        return cls()

class B(A):
    @classmethod
    def f1(cls) -> typing.Self:
        return super(B, cls).f1()

    def f2(self):
        pass

B.f1().f2()
```

Results in:
```
E1101: Instance of 'A' has no 'f2' member; maybe 'f1'? (no-member)
```

mypy does work correctly.

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
a.py:1:0: C0114: Missing module docstring (missing-module-docstring)
a.py:3:0: C0115: Missing class docstring (missing-class-docstring)
a.py:5:4: C0116: Missing function or method docstring (missing-function-docstring)
a.py:3:0: R0903: Too few public methods (1/2) (too-few-public-methods)
a.py:8:0: C0115: Missing class docstring (missing-class-docstring)
a.py:13:4: C0116: Missing function or method docstring (missing-function-docstring)
a.py:16:0: E1101: Instance of 'A' has no 'f2' member; maybe 'f1'? (no-member)
```

### Expected behavior

The return value of `B.f1()` is `B` and not `A`.

### Pylint version

```shell
pylint 3.0.1
astroid 3.0.1
Python 3.11.6 (tags/v3.11.6:8b6ee5b, Oct  2 2023, 14:57:12) [MSC v.1935 64 bit (AMD64)]
```

### OS / Environment

*

**Comment by Pierre-Sassoulas** [comment]
Fixed in 4.0.0
